lib/biokbase/catalog/db.py

```diff
@@ -120,18 +122,75 @@ class MongoCatalogDBI:
 
     def __init__(self, mongo_host, mongo_db, mongo_user, mongo_psswd, mongo_authMechanism):
 
-        # create the client
-        self.mongo = MongoClient('mongodb://' + mongo_host)
+        self.mongo_host = mongo_host
+        self.mongo_db = mongo_db
+        self.mongo_user = mongo_user
+        self.mongo_psswd = mongo_psswd
+        self.mongo_authMechanism = mongo_authMechanism
+
+        self.lock = threading.Lock()
+        self.mongo_client = None
+        self.client_initialized = False
+
+        # Initialize mongo client
+        self._initialize_mongo_client()
+
+        # Create dbs and collections
+        self._create_collections()
+
+        # Check the db schema
+        self.check_db_schema()
+
+        # Create indexes
+        self._create_indexes()
+
+        # Close the MongoDB client manually before forking
+        self._close_mongo_client()
+
+    def _initialize_mongo_client(self):
+        """Initialize MongoDB client with lock to prevent race conditions."""
+        # Use the lock to ensure only one thread initializes the mongo client at a time
+        with self.lock:
+            try:
+                if not self.client_initialized or self.mongo_client is None:
```

Comment:
```suggestion
        """Initialize MongoDB client with lock to prevent race conditions."""
        # Use the lock to ensure only one thread initializes the mongo client at a time
        if self.mongo_client:
            return
        with self.lock:
            try:
                if not self.client_initialized or self.mongo_client is None:
```

Don't even pay the cost of entering the lock if the client is available

Reply:
👍

Reply:
:+1: